### ISSUE TITLE
NextBot.OnSight & NextBot.OnLostSight

### ISSUE THREAD

**Issue by Dragoteryx** [Facepunch/garrysmod-requests#1649]
With all those nextbot sight functions added to the game (thank you Rubat), I tried setting `ng_debug` to `VISION` to see if it worked, and I was surprised to find that it actually worked. (I don't think it used to before)
![It works!](https://media.discordapp.net/attachments/418480949542780938/729481523153207317/unknown.png)

Would it be possible to add two new hooks for nextbots called NextBot.OnSight & NextBot.OnLostSight?

Even though NextBot.IsAbleToSee works on any entity, it seems this code only checks for line of sight with players and other nextbots. (adding support for "classic" NPCs would be nice if possible)

**Comment by Dragoteryx** [comment]
I ran some tests with `nb_debug VISION` and noticed some odd behaviour, mostly `OnSight` not being called when it should have been, so I checked the C++ code and noticed that the nextbot vision system is way more complex than I anticipated.

I might (and probably am) be wrong but this is why that happens:
When a nextbot sees an entity it already knew about, `OnSight` is called. However, when the nextbot sees an entity for the first time, that entity is marked as seen, but `OnSight` isn't called as the nextbot didn't know about this entity. Also, the nextbot forgets seen entities after a while if they leave the nextbot's sight for a while.
